pasture-algorithms/src/voxel_grid.rs

```diff
@@ -0,0 +1,875 @@
+            &centroid.set_attribute(&attributes::EDGE_OF_FLIGHT_LINE, mc_slice);
+        } else if &a.name() == &attributes::CLASSIFICATION.name() {
+            let most_common =
+                centroid_most_common::<PB>(v, buffer, &attributes::CLASSIFICATION, a.datatype())
+                    as u8;
+            let mc_slice = unsafe { view_raw_bytes(&most_common) };
+            &centroid.set_attribute(&attributes::CLASSIFICATION, mc_slice);
+        } else if &a.name() == &attributes::SCAN_ANGLE_RANK.name() {
+            let most_common =
+                centroid_most_common::<PB>(v, buffer, &attributes::SCAN_ANGLE_RANK, a.datatype())
+                    as i8;
+            let mc_slice = unsafe { view_raw_bytes(&most_common) };
+            &centroid.set_attribute(&attributes::SCAN_ANGLE_RANK, mc_slice);
+        } else if &a.name() == &attributes::SCAN_ANGLE.name() {
+            let most_common =
+                centroid_most_common::<PB>(v, buffer, &attributes::SCAN_ANGLE, a.datatype()) as i16;
+            let mc_slice = unsafe { view_raw_bytes(&most_common) };
+            &centroid.set_attribute(&attributes::SCAN_ANGLE, mc_slice);
+        } else if &a.name() == &attributes::USER_DATA.name() {
+            let most_common =
+                centroid_most_common::<PB>(v, buffer, &attributes::USER_DATA, a.datatype()) as u8;
+            let mc_slice = unsafe { view_raw_bytes(&most_common) };
+            &centroid.set_attribute(&attributes::USER_DATA, mc_slice);
+        } else if &a.name() == &attributes::POINT_SOURCE_ID.name() {
+            let most_common =
+                centroid_most_common::<PB>(v, buffer, &attributes::POINT_SOURCE_ID, a.datatype())
+                    as u16;
+            let mc_slice = unsafe { view_raw_bytes(&most_common) };
+            &centroid.set_attribute(&attributes::POINT_SOURCE_ID, mc_slice);
+        } else if &a.name() == &attributes::COLOR_RGB.name() {
+            let color = centroid_average_vec(v, buffer, &attributes::COLOR_RGB, a.datatype());
+            let color_u16 = Vector3::new(color.x as u16, color.y as u16, color.z as u16);
+            let col_slice = unsafe { view_raw_bytes(&color_u16) };
+            &centroid.set_attribute(&attributes::COLOR_RGB, col_slice);
+        } else if &a.name() == &attributes::GPS_TIME.name() {
+            let max =
+                centroid_max_pool::<PB>(v, buffer, &attributes::GPS_TIME, a.datatype()) as f64;
+            let m_slice = unsafe { view_raw_bytes(&max) };
+            &centroid.set_attribute(&attributes::GPS_TIME, m_slice);
+        } else if &a.name() == &attributes::NIR.name() {
+            let average =
+                centroid_average_num::<PB>(v, buffer, &attributes::NIR, a.datatype()) as u16;
+            let avg_slice = unsafe { view_raw_bytes(&average) };
+            &centroid.set_attribute(&attributes::NIR, avg_slice);
+        } else if &a.name() == &attributes::WAVE_PACKET_DESCRIPTOR_INDEX.name() {
```

Comment:
I think the waveform data makes no sense in this algorithm. You can comment it out, maybe there will be a use-case for it in the future, but for now we can just have the method check that the layout of the `filtered_buffer` does not contain any of the waveform values. 

A future version of this algorithm should take a separate object that contains a mapping between point attributes and the desired type of 'reduction function'. This way we could store defaults in this object but have the users overwrite the defaults with whatever they want.